Added tree comparison

Ethrel opened the pull request:
Adding tree comparison

Relates to #2141

Realized it was just LUA, so it wasn't a huge deal to just throw my hat in the ring for a minute.

Quotae commented:
This is a great first PR, very useful feature. Would like to see a few changes though:
1. Would be nice if this extra tree dropdown disappeared when you're not comparing them, instead of being grayed out. The node power buttons disappear when not in use, check out how they do it.
![image](https://user-images.githubusercontent.com/39030429/107309936-0264fc80-6a51-11eb-8084-068b86d7bc9b.png)

2. Spacing between the Compare: text and the button to the left of it is a bit off versus others. See:
Yours: ![image](https://user-images.githubusercontent.com/39030429/107310082-4ce67900-6a51-11eb-90cc-bed1cfacee92.png)
Others: ![image](https://user-images.githubusercontent.com/39030429/107310118-5f60b280-6a51-11eb-9266-ceba00a63ff3.png)

3. Issue: if you are comparing trees and also showing node power, the colors of the tree comparison don't work correctly:
![image](https://user-images.githubusercontent.com/39030429/107310237-a6e73e80-6a51-11eb-9529-d30b1db25d63.png)

4. Trying to compare trees of two different game versions causes a crash. Here's a build with a 3.12 tree in case you don't have one: https://pastebin.com/GBYMZDDV Just open the build, head to the tree tab, click "Convert to 3.13" at the bottom, and try to compare one of the 3.12 trees to the 3.13 one.

Ethrel commented:
Ok. So, I updated the commit, squashed it all down, we're now all nice and happy in a single commit.

Hiding the dropdown was a little more complicated than the node power version. The node power version has the advantage of being able to just hide everything anchored to it. In my case, that couldn't be done like that, so I had to do some things (See https://github.com/PathOfBuildingCommunity/PathOfBuilding/pull/2151/commits/221ca7f0b950a7756c41395dbc69f19660ceb091#diff-5fcc1cbe087d23273aa0779a6e056d56b04c07acef5ce3f943b9cd402048353fR74). Hopefully that's how it's meant to be done rather than just some hack.

ppoelzl commented:
Please no unrelated changes to the formatting. This makes it harder to review than necesary.

Wires77 commented:
I get this error in the Items tab when hovering over a socketed jewel.  Using the pastebin that @Quotae posted, even with the tree comparison off.

![image](https://user-images.githubusercontent.com/1209372/107551776-4a426b80-6b98-11eb-8028-3462dc4c0fd1.png)

Ethrel commented:
Didn't even realize that the jewel's tree position was drawn on jewel hover. Just a method call that didn't match the new signature. Fixed that now too :D

Wires77 commented:
> 
> 
> Didn't even realize that the jewel's tree position was drawn on jewel hover. Just a method call that didn't match the new signature. Fixed that now too :D

Oh!  Then I'm guessing...yep, this errors out too in the Calcs tab:

![image](https://user-images.githubusercontent.com/1209372/107557016-b2944b80-6b9e-11eb-96f8-149d79245081.png)

I'd run a file search for the method you changed, in case there are other places I can't think of right now

Ethrel commented:
Okay. So when you pointed out another place, I did indeed do a full source search. Found the one for the calcs tab with my search, and that seems to be the only one left. And that one method was the only one that wasn't a function defined within a local scope of somewhere, so it should be the only breakpoint for the same incorrect method arguments call issue.

Wires77 commented on Classes/PassiveTreeView.lua:
```diff
@@ -93,10 +93,12 @@ function PassiveTreeViewClass:Save(xml)
 	local spec = build.spec
 	local tree = spec.tree
 
+	local isComparing = compareSpec and true or false;
```
This can be simplified in `PassiveTreeView` by removing the `isComparing` boolean and just checking `compareSpec` everywhere instead.  They can be substituted 1-for-1

```suggestion
```

Wires77 commented on Classes/PassiveTreeView.lua:
```diff
@@ -93,10 +93,12 @@ function PassiveTreeViewClass:Save(xml)
 end
 
-function PassiveTreeViewClass:Draw(build, viewPort, inputEvents)
+function PassiveTreeViewClass:Draw(build, compareSpec, viewPort, inputEvents)
```
After looking at the code around other `Draw` functions, none of the others have data this specific in them.  They only pass in the viewPort most of the time, along with `inputEvents`.  Most of the time we pass the value needed to the `PassiveTreeView` from the calling class.  E.g. in `TreeTab.lua` you could do `self.viewer.compareSpec = compareSpec` instead of passing it to this function.  See `showHeatMap` for an example, too.

If we could refactor this slightly to follow the existing convention that'd be great.  Then you don't have to insert `nil` as part of the method call in the other tabs (Sorry, I looked at the behavior before the code, otherwise I would've suggested this before you made those changes!)

Wires77 reviewed (changes requested):
The functionality works great, I finally got a chance to look at the code itself and made some suggestions.  Once those are addressed this looks good to go!

Wires77 reviewed (approved):
LGTM
